File: typescript/systemdesign/inMemoryFileSystem.ts
/*
In-memory file system supporting mkdir, addFile/write, ls, and readFile
over a tree of directory and file nodes, addressed by absolute Unix-style
paths ("/a/b/c").

status - completed
*/

import { logp } from "../utils/logger";

type FsNode = DirectoryNode | FileNode;

class DirectoryNode {
  readonly type = "directory" as const;
  readonly children: Map<string, FsNode> = new Map();
}

class FileNode {
  readonly type = "file" as const;
  content = "";
}

function splitPath(path: string): string[] {
  return path.split("/").filter((segment) => segment.length > 0);
}

class InMemoryFileSystem {
  private readonly root = new DirectoryNode();

  private resolveDir(segments: string[], createMissing: boolean): DirectoryNode {
    let current = this.root;
    for (const segment of segments) {
      let next = current.children.get(segment);
      if (!next) {
        if (!createMissing) {
          throw new Error(`no such directory: ${segment}`);
        }
        next = new DirectoryNode();
        current.children.set(segment, next);
      }
      if (next.type !== "directory") {
        throw new Error(`not a directory: ${segment}`);
      }
      current = next;
    }
    return current;
  }

  mkdir(path: string): void {
    this.resolveDir(splitPath(path), true);
  }

  writeFile(path: string, content: string): void {
    const segments = splitPath(path);
    const fileName = segments.pop();
    if (!fileName) {
      throw new Error("path must include a file name");
    }
    const dir = this.resolveDir(segments, true);
    const existing = dir.children.get(fileName);
    if (existing && existing.type !== "file") {
      throw new Error(`not a file: ${fileName}`);
    }
    const file = existing ?? new FileNode();
    file.content = content;
    dir.children.set(fileName, file);
  }

  readFile(path: string): string {
    const segments = splitPath(path);
    const fileName = segments.pop();
    if (!fileName) {
      throw new Error("path must include a file name");
    }
    const dir = this.resolveDir(segments, false);
    const node = dir.children.get(fileName);
    if (!node || node.type !== "file") {
      throw new Error(`no such file: ${path}`);
    }
    return node.content;
  }

  ls(path: string): string[] {
    const dir = this.resolveDir(splitPath(path), false);
    return Array.from(dir.children.keys()).sort();
  }
}

// usage scenarios

const fs = new InMemoryFileSystem();
fs.mkdir("/home/user/docs");
fs.writeFile("/home/user/docs/notes.txt", "first draft");
fs.writeFile("/home/user/todo.txt", "buy milk");
fs.mkdir("/home/user/docs/archive");

logp(`ls("/home/user") -> ${JSON.stringify(fs.ls("/home/user"))}`);
logp(`ls("/home/user/docs") -> ${JSON.stringify(fs.ls("/home/user/docs"))}`);
logp(`readFile("/home/user/docs/notes.txt") -> "${fs.readFile("/home/user/docs/notes.txt")}"`);

fs.writeFile("/home/user/docs/notes.txt", "revised draft");
logp(`readFile after overwrite -> "${fs.readFile("/home/user/docs/notes.txt")}"`);

try {
  fs.readFile("/home/user/docs/missing.txt");
} catch (err) {
  logp(`readFile on missing file threw as expected: ${(err as Error).message}`);
}

try {
  fs.mkdir("/home/user/todo.txt/nested");
} catch (err) {
  logp(`mkdir under a file threw as expected: ${(err as Error).message}`);
}

File: python/systemdesign/in_memory_file_system.py
"""
In-memory file system supporting mkdir, write file, ls, and read,
backed by a tree of directory nodes each holding child directories
and files, similar to the classic "design in-memory file system"
interview problem.

status - completed
"""

from base_logger.logging_event import create_logger


class DirectoryNode:

    def __init__(self):
        self.directories = {}
        self.files = {}


class InMemoryFileSystem:

    log = create_logger(__name__)

    def __init__(self):
        self.root = DirectoryNode()

    def _split(self, path):
        return [part for part in path.strip("/").split("/") if part]

    def _resolve_dir(self, parts, create=False):
        node = self.root
        for part in parts:
            if part not in node.directories:
                if not create:
                    raise FileNotFoundError(f"directory not found: {part}")
                node.directories[part] = DirectoryNode()
            node = node.directories[part]
        return node

    def mkdir(self, path):
        parts = self._split(path)
        self._resolve_dir(parts, create=True)

    def write_file(self, path, content):
        parts = self._split(path)
        if not parts:
            raise ValueError("path must include a file name")
        *dir_parts, file_name = parts
        directory = self._resolve_dir(dir_parts, create=True)
        directory.files[file_name] = content

    def read_file(self, path):
        parts = self._split(path)
        *dir_parts, file_name = parts
        directory = self._resolve_dir(dir_parts, create=False)
        if file_name not in directory.files:
            raise FileNotFoundError(f"file not found: {path}")
        return directory.files[file_name]

    def ls(self, path):
        parts = self._split(path)
        if parts and parts[-1] in self._resolve_dir(parts[:-1], create=False).files:
            return [parts[-1]]
        directory = self._resolve_dir(parts, create=False)
        return sorted(list(directory.directories.keys()) + list(directory.files.keys()))


if __name__ == "__main__":
    log = create_logger("in_memory_fs_demo")

    fs = InMemoryFileSystem()
    fs.mkdir("/a/b/c")
    fs.write_file("/a/b/c/hello.txt", "hello world")
    fs.write_file("/a/readme.md", "top level readme")

    log.info("ls('/a') -> %s", fs.ls("/a"))
    log.info("ls('/a/b/c') -> %s", fs.ls("/a/b/c"))
    log.info("read('/a/b/c/hello.txt') -> %s", fs.read_file("/a/b/c/hello.txt"))
    log.info("ls('/a/readme.md') -> %s (ls on a file returns just that file)",
             fs.ls("/a/readme.md"))

    try:
        fs.read_file("/a/b/c/missing.txt")
    except FileNotFoundError as exc:
        log.info("reading missing file raised: %s", exc)

    try:
        fs.read_file("/no/such/dir/file.txt")
    except FileNotFoundError as exc:
        log.info("reading from missing directory raised: %s", exc)
